Improve alphabetic sort, refs #5173

djjuhasz opened the pull request:
- Add custom 'alphasort' normalizer that converts text to lowercase, removes punctation, and does ASCII folding
- Add "sortFields" parameter to ES mapping which creates special "alphasort" keyword field
- Add information_object_i18n.title and actor.authorizedFormOfName "alphasort" fields to ES index
- Use alphasort field for actor, information object, and repository sorting

djjuhasz committed:
Improve alphabetic sort, refs #5173

- Add custom 'alphasort' normalizer
- Apply 'alphasort' normalizer to special 'alphasort' ES field
- Add alphasort fields for accession, actor, information object,
  repository, term
- Use alphasort fields for ES sorting where appropriate

djjuhasz commented:
Note: More sort fields can be added, but I wanted to get eyes on the code before adding too many other resource types.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchMapping.class.php:
```diff
@@ -482,4 +484,27 @@ protected function getNestedI18nObjects($languages, $nestedI18nFields)
+   * Add i18n "sort" field that is not tokenized, lowercase, has whitespace
+   * trimmed, punctation removed, and is ASCII folded for more natural sorting
+   */
+  protected static function addSortFields(&$nestedI18nFields, $typeProperties)
```
Why static?

djjuhasz replied:
@jraddaoui just because neither of the input variables are stored in the object context.  I lean towards static methods by default, unless there is a need to get or set the object state.

djjuhasz replied:
Looks like I might be "doing it wrong" though 😛 
https://stackoverflow.com/questions/33705976/when-should-i-use-static-methods

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchMapping.class.php:
```diff
@@ -482,4 +484,27 @@ protected function getNestedI18nObjects($languages, $nestedI18nFields)
+        'type' => 'keyword',
+        'normalizer' => 'alphasort',
+        'include_in_all' => false,
+        'store' => false);
```
I think `store` is set to `false` by default.

djjuhasz replied:
@jraddaoui yes, looks like `store` defaults to `false` for the keyword datatype: https://www.elastic.co/guide/en/elasticsearch/reference/5.6/keyword.html

I'll update the PR

jraddaoui reviewed (commented):
This is looking amazing, thanks @djjuhasz!

jraddaoui commented:
I didn't approve the PR right away because I thought you'll be adding more changes, but let me know if that's not the case ;)

djjuhasz commented:
> I didn't approve the PR right away because I thought you'll be adding more changes, but let me know if that's not the case ;)

You're right @jraddaoui, my plan is to add sort fields for more resource types (terms for sure, maybe accessions?)

djjuhasz commented:
Okay, I've rolled up all the commits and updated the commit message.  I think this is ready to merge.

jraddaoui reviewed (approved):
Awesome, thanks @djjuhasz!!

Please, try not to squash until the review is finished. Not a big deal in this case, but it can complicate the process for the reviewer when the PR involves a lot of changes.

djjuhasz commented:
@jraddaoui okay, I will avoid squashing dev branches.

Just FYI in the past I've gotten comments when I do *not* squash my branch for a PR to "make sure I squash the merge commit and update the commit message."  I squashed the branch and updated the commit message this time to be clear *exactly* what I want to merge.

jraddaoui commented:
Thanks @djjuhasz!

Yes, please squash and clean the history and commit message before merging, but try not to do it in the middle of a review, just at the end, and maybe ask for another review to check the squash if you want (which shouldn't involve re-reading the changes again, just the commit history and message).

This is just my opinion and it depends on the PR. In a minor PR it won't impact that much the review time, but it can really affect in big PRs.
